Remove unwanted umask

PierreRambaud opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | We don't want to force the 0777 mode. The default must be 0755 for directories or 0644 for files. Change `getcwd()` by `__DIR__`.
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixes? | Fixes #12108
| How to test?  | Add product images, create thumbnails, regenerate robots.txt and htaccess. Install modules, try to do everything around file creation and update. I will try to test this under Windows env as soon as possible.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/12124)
<!-- Reviewable:end -->

Commit:
Remove unwanted getcwd and replace by __DIR__

Commit:
Remove unwanted chmod and change default umask by 0755

rdy4ever commented:
I think this should be in 1.7.5.1

PierreRambaud commented:
This is not a regression from the latest release. WDYT @eternoendless @matks

rdy4ever commented:
Yes, it is not a regression but it's an important security enhancement that fixes something that shouldn't have been present in 1.7.5 to begin with.

PierreRambaud commented:
I understand what you mean but I'm not the one who take this decision :smiley:

eternoendless commented:
> [This] fixes something that shouldn't have been present in 1.7.5 to begin with.

While I agree with that, I think it's too risky tu push in a patch version. I don't think it's critical either, so to me it sounds more reasonable to push it into `develop`

matks commented:
> > [This] fixes something that shouldn't have been present in 1.7.5 to begin with.
> 
> While I agree with that, I think it's too risky tu push in a patch version. I don't think it's critical either, so to me it sounds more reasonable to push it into `develop`

+1
People expect a patch to be totally harmless.

matks commented on admin-dev/ajax-tab.php:
```diff
@@ -23,11 +23,8 @@
-
 if (!defined('_PS_ADMIN_DIR_')) {
-    define('_PS_ADMIN_DIR_', getcwd());
+    define('_PS_ADMIN_DIR_', __DIR__);
```
This seems to be out of the scope of this PR ?

PierreRambaud replied:
Yes, but it's a bug :/
I forgot to update the description ^^

matks replied:
I'm worried about the impact of this change for some server configuration. Like shops in sub-folders.
Shall we require extensive QA tests ?

PierreRambaud replied:
Don't know how it can have an impact, there is a huge difference between cwd & dir.
If you are running the script in other directory than the admin, it failed :)

jolelievre replied:
Agreed with @PierreRambaud `getcwd` return the execution path not the file path, which could have really bad side effects in this case. However in our usual web context where they end up having the same value it is safe to have this modification.
I don't think a thorough QA needs to be done, simply check each modified end point, if something is broken it should break instantly anyway.

matks commented on classes/PrestaShopAutoload.php:
```diff
@@ -260,7 +260,7 @@ public function dumpFile($filename, $content)
         }
         // Ignore for filesystems that do not support umask
-        @chmod($tmpFile, 0666);
+        @chmod($tmpFile, file_exists($filename) ? fileperms($filename) : 0666 & ~umask());
```
For someone not used to bitwise operators (like me 😛 ) the meaning of `& ~` is not obvious.
Do you think we could name a variable to make it more obvious ?

PierreRambaud replied:
This function is a copy of the Symfony autoloader, I just update the behavior to keep the filepermissions and get umask configuration if someone change it. (Like Symfony does)

PierreRambaud replied:
For the variable name, it is overkill :/

jolelievre replied:
I also think a variable is overkill, and it won't help much to understand the bit operations if you don't know them ^^

matks commented on controllers/admin/AdminCartsController.php:
```diff
@@ -427,8 +427,6 @@ public function ajaxProcessUpdateCustomizationFields()
                             continue;
                         } elseif (!ImageManager::resize($tmp_name, _PS_UPLOAD_DIR_ . $file_name . '_small', (int) Configuration::get('PS_PRODUCT_PICTURE_WIDTH'), (int) Configuration::get('PS_PRODUCT_PICTURE_HEIGHT'))) {
                             $errors[] = $this->trans('An error occurred during the image upload process.', array(), 'Admin.Catalog.Notification');
-                        } elseif (!chmod(_PS_UPLOAD_DIR_ . $file_name, 0777) || !chmod(_PS_UPLOAD_DIR_ . $file_name . '_small', 0777)) {
```
It is relevant to remove this processing however can we replace it by an harmless permissions check ? I think this line has been added to handle a bad permissions settings of `_PS_UPLOAD_DIR_`, maybe we can keep the warning ?

PierreRambaud replied:
It is not a warning, it a filemode change, nothing more.

jolelievre replied:
I agree with mathieu, it would be nice to keep a check on these permissions, using `fileperms` and bit operations to check the permissions Something like `fileperms($filename) & FileSystem::DEFAULT_MODE == FileSystem::DEFAULT_MODE`

PierreRambaud replied:
There is actually no check and no warning, it is an error.. chmod is not like fileperms.
More, we need to don't force the file change, if someone change is own filesystem mode we mustn't add an error each time he cames here.

jolelievre replied:
Yes I understand the difference between `chmod` and `fileperms`, I'm well aware that it wouldn't be the same behavior
But the thing is if the minimum permissions are not set I'm afraid the feature won't work correctly because the image won't be writable, then maybe we should return an explicit error about file permissions so that the user understands that he needs to configure it right

PierreRambaud replied:
We already have this kind of check during installation / upgrade.
If you change the files/directories permissions after because of security or anything else, you don't want to have an error while using an ajax request because of this.
The resize has already been done, so the file is already created.

jolelievre replied:
Ok I'm convinced But then I wonder why this chmod was performed before?

PierreRambaud replied:
To force the 0755 mode only :smile:

matks commented on controllers/front/ProductController.php:
```diff
@@ -825,8 +825,6 @@ protected function pictureUpload()
                     $this->errors[] = $this->trans('An error occurred during the image upload process.', array(), 'Shop.Notifications.Error');
                 } elseif (!ImageManager::resize($tmp_name, _PS_UPLOAD_DIR_ . $file_name . '_small', $product_picture_width, $product_picture_height)) {
                     $this->errors[] = $this->trans('An error occurred during the image upload process.', array(), 'Shop.Notifications.Error');
-                } elseif (!chmod(_PS_UPLOAD_DIR_ . $file_name, 0777) || !chmod(_PS_UPLOAD_DIR_ . $file_name . '_small', 0777)) {
```
Same for here 😉 remove the processing, keep the warning ?

jolelievre replied:
same check as above with `fileperms`

mickaelandrieu commented:
IMHO this will create more bugs than it fixes :/

PierreRambaud commented:
@mickaelandrieu Don't understand what kind of bug this PR will create? It is a security issue to force 0777..

mickaelandrieu commented:
and people will then complain they can't access to their files while being totally unable to change the rights (they're not developers and/or they don't have the access to a CLI entry point).

We did it for a good reason, we need to search for in the history why :+1:

PierreRambaud commented:
So let's keep the old controllers, and objects models, they did it for a good reason too :wink:

It is a real security issue, it breaks dedicated and hosted server security. I don't want all user from my server to read what code I host :/
We must focus on documentation and help people to understand how to configure their environment when needed instead injected a really bad practice.
Otherwise, why we don't keep PHP5.2 for people who don't know how to update their server?

eternoendless commented:
If they are not developers, then maybe they shouldn't be messing around with the files easily? They risk breaking their shop.

If they need to change files created by PrestaShop, and if they don't have su access, they then can ask their system administrator to do it for them. This could be added to a Troubleshooting section in the [User Guide](http://doc.prestashop.com/display/PS17/User+Guide). 

Wdyt @colinegin? Do we risk making shops insecure by default or make it easier for merchants to mess around with the files?

mickaelandrieu commented:
> We did it for a good reason, we need to search for in the history why +1

I agree we should merge this if you guys take responsibility for every new "issues" spotted by people impacted by this change.

Go or no Go? I'll ping you too @markaussie of course! :angel:

markaussie commented:
> > We did it for a good reason, we need to search for in the history why +1
> 
> I agree we should merge this if you guys take responsibility for every new "issues" spotted by people impacted by this change.
> 
> Go or no Go? I'll ping you too @markaussie of course! 👼

Hi,
The reason for the change is still unanswered?

PierreRambaud commented:
If they need su access to specify something, this is something apache or nginx must not do.. 
This is why all good operating systems have acl. 

Don't forget it is a security issue. PHP scripts run on the weberver. Leaving this kind of permissions that way will make your webserver user (www-data or apache by default) able to write on those files.
In case of you add a vulnerability on other service running on the server, you allow everyone to:

* Loss of everything (the write permission is also permission to delete the files);
* Addition of unwanted stuff: some attacks aim to add data to your site.

So, yeah, it's a horrible idea to leave permissions on that state.

As it said by @eternoendless if they encountered issue with their configuration, this could be added to a Troubleshooting section in the [User Guide](http://doc.prestashop.com/display/PS17/User+Guide).

colinegin commented:
Hi all,

Sorry for this late reply ! 
In my opinion, the safest the best :) 
I mean, it's our role to prevent as much as we can people (especially non technical ones) from getting security flaws or suffer global loss. 

So I agree with @PierreRambaud, we should not leave permissions on that state, and make it more restricted by default. In the meantime, I think we should explain how to bypass this restriction so that developers and/or agencies who want the previous behaviour can stay 100% free while handling their files.

Just a thought, is there any potential risk for merchants who upgrade their store ? At least they should get informed that they will no longer have 0777 rights on their directories and files.

What do you guys think ?

julianmatz commented:
Hi all. After reading through the comments here, I just wanted to add, in case anyone wasn't aware, this issue **forces** users to have 0666/0777 permissions. I had my file permissions set up correctly. I then upgraded a Prestashop installation and suddenly had o+w permissions. And I would not have noticed this except that I have a security daemon running on my server that alerts me of these things. After manually running chmod o-w on my whole file tree, the permissions automatically reverted back to o+w. It took me some time to figure out why.  So, I'd personally be more inclined to ask if there is a potential risk for merchants affected by this issue currently, perhaps without even knowing.

PierreRambaud commented:
ping @PrestaShop/prestashop-core-developers

jolelievre commented on src/Core/Foundation/Filesystem/FileSystem.php:
```diff
@@ -30,6 +30,11 @@
+    /**
+     * Default mode for directories and files
+     */
+    const DEFAULT_MODE = 0755;
```
The default mode for files and folder should not be the same:
- folder: 755
- files: 644

So you should have two different const `FOLDER_DEFAULT_MODE` and `FILE_DEFAULT_MODE` (even though you used this const only once for a file It's safer to differentiate both modes).

Besides, depending on your server config sometimes you need to have more open permissions, I'm thinking 775 and 664 permissions (for cases where your server user doesn't match your cli user for example). In which case it is convenient to have a common group and allow writing for this group.

So maybe this modes should be editable in the back office and we could use static getters in this class:

```
public static function getFolderDefaultMode() {
    return Configuration::get('PS_FOLDER_DEFAULT_MODE', self::FOLDER_DEFAULT_MODE);
}

public static function getFileDefaultMode() {
    return Configuration::get('PS_FILE_DEFAULT_MODE', self::FILE_DEFAULT_MODE);
}
```

Maybe it's overkill for the scope of this PR but I think it would be really nice to have and it would prevent us to break some customers environment by allowing them to override this modification.

PierreRambaud replied:
You're right, I add the change for 0644 ;) But it's a little overkill to add this in configuration, more because many developers don't know how it works, so it'll add a real complexity for merchants.

jolelievre replied:
Well if they don't know how it works they won't look for the setting anyway :P
But if they are stuck because of our modif we would have at least a way to help them fix the issue through the admin
I thought adding these two fields in `Advanced Parameters -> Administration` wouldn't be that long to implement, but it could really save us time for support later

jolelievre replied:
Everything else seems fine, this is the last feedback to deal

As I read the previous conversation I saw @colinegin also proposed to have a way to bypass this restriction. With the current modification the only possibility is to  modify the const in `Filesystem` (which is not overridable), that's not great approach.

Adding these two configurations would be useful for tricky environment when, unfortunately, you can't manage your permissions yourself, or when multiple users try to access/edit these files.

PierreRambaud replied:
If they don't know how it works, they will touch and break all the thing because they will try to edit something to try to understand how it works.
This is a server configuration problem, if you need to have multiple users to play with your files, you are maybe on linux in development mode and you can configure your apache (**/etc/apache2/envars** for example) to use your own user/group instead **www-data**.

More, all FTP tools allow you to configure permissions and if you are not using an FTP, this is because you know how to configure a server, so, you should be able to set permissions and the new code will not override it.

We must not add system configuration, even wordpress has a good article about that: https://codex.wordpress.org/Changing_File_Permissions

jolelievre replied:
Yeas I know you can change the group from apache. That's the kind of usecase I was thinking about. But in that case to allow write permissions between a group, should we change the default permission to 775 and 664?

PierreRambaud replied:
If you have the same group yes, but the file will be created with the running apache user. And it becomes a complicated configuration which is not a good practice (not at all).

jolelievre replied:
ok then, let's keep it simple for now

jolelievre reviewed (changes requested):
Great work! this is a very good idea to clean this! I had a few comments nonetheless
And there is a remaining conflict

Commit:
Separate files and directories mode

- use DEFAULT_MODE_FOLDER (0755) for folders
- use DEFAULT_MODE_FILE (0644) for files

jolelievre reviewed (approved):
nice job!

matks commented:
Thanks @PierreRambaud and everybody who participated to this discussion

rGaillard commented:
@PierreRambaud doing this, people will use the same user for PHP and FTP/SSH or try to put the whole website in 0777 because they will not be able to delete / change files directly. And then PHP will be able to write in the whole website and not only in the upload folders (where php execution is disabled).

julianmatz commented:
@rGaillard If they do this, they will have an insecure ecommerce store. But it will be on the users and not the developers. It is the reason people normally consult professionals or experts. Also, there should be at least a number of alternative ways to address the issue - for example, if you want to umask, then maybe you could do it conditionally based on some configuration flag.

PierreRambaud commented:
Totally agreed with @julianmatz , using 077 is maybe the worst thing to do. But they can configure www-data to have the same access as their FTP/SSH users. More if they use ssh, they should be able to do a chown / chmod.

And, if you have a security issue in an other service host on the same server, you are able to write a php and do everything you want because of 0777.

rGaillard commented:
> Totally agreed with @julianmatz , using 077 is maybe the worst thing to do. But they can configure www-data to have the same access as their FTP/SSH users. More if they use ssh, they should be able to do a chown / chmod.
> 

This is the problem. PHP will be able to write anywhere on the website and not only where the PHP execution is disabled. The risk is higher here! And with this PR, this is what people will do.

rGaillard commented:
> And, if you have a security issue in an other service host on the same server, you are able to write a php and do everything you want because of 0777.

There is more chance that there is an issue on the website (specific development, modules etc...) than that there is no chroot or open_basedir configured for the website. Most users don't know how to manage rights, so, they will all put on 0777 or use the same PHP user for the whole website because of this new configuration.

julianmatz commented:
I don't know about "most" users. I think if they don't know how to set correct permissions, they might also have trouble creating a DB and DB user, changing permissions in the first place, and definitely changing the user running PHP. But I get it, you want to make it easy for users. I'm just saying, there's other ways to do this. If they can be guided through an installation and guided through a DB wizard, they could be guided through this also. Trading off security for simplicity isn't a good idea.

PierreRambaud commented:
There were already many 077 in the code, this is why this PR exists, even Wordpress have a good article about how to configure and set permissions: https://codex.wordpress.org/Changing_File_Permissions

For people who use FTP or mutual host, they have already a specific user which is the same as the Apache (for all hosts providers).
For custom installation / dedicated server, it's maybe the worst thing to do, if they don't know how to set permissions, or how to configure a Linux environnement, they need to learn or we need to write a doc for that.
As @julianmatz says, setting 0777 or forcing the umask is a really really bad practice..

markaussie commented:
Well done.
Suggestion: Add to the BO->Information page a list of the websites folder and file permissions with warnings for insecure settings and green ticks when things are set securely.
Enhanced with option to toggle settings.

Security should be a priority, transparent and part of The Future Architecture.

MathiasReker commented:
I just made a module to fix file- and folder permissions: https://github.com/MathiasReker/filepermissions :-)

MathiasReker commented:
644 for files, 755 for folders.

Any folders with 777? cache folder, or img folder?

ghost commented:
you prefer for security to permits to a new vendor who's starting without developper to insecure his shop and permissions to all hacker and all kind of problem to acces freely rewrite files htacess etc ... and sure touch the integrity of online payment with papaya official for example so you prefer for security to permisss insecurity and so destroy a shop and the person who have the shop

ghost commented:
@mickaelandrieu @matks  ANd so when we update to next version we must do it again ! is that 

En français au cas car il y en a beaucoup de marchands désespérés à ce sujet et qui ne trouvent pas la solution et ne savent pas quoi faire 

Il faut donc modifier : 
classes/PrestaShopAutoload.php
controllers/admin/AdminCartsController.php
controllers/front/ProductController.php
src/Core/Foundation/Filesystem/FileSystem.php Outdated
admin-dev/ajax-tab.php

Et donc quand on fera la prochaine update par exemple vers 1.7.6 on devra donc recommencer le tout car ce ne sera pas modifier c'est bien cela !? 


Donc vous préférez par sécurité, permettre aux hackers, a toutes les personnes mal intentionnées de pouvoir modifier une boutique prestashop lancée  par un marchands débutant sans savoir ce problème de sécurité et de permissions car combien ne sont pas au courant ? et donc mettre à mal ainsi le vendeur et sa boutique car cela touche tous les modules et fichier y compris les moyens de paiement ..  parce par sécurité prestashop à préférer permettre ce type de procédure non sécurisé ! 

Merci beaucoup Prestashop ! et merci pour laisser cette insécurité totale s'installée et permettre le hacking d'une boutique comme nous venons de le subir parce que prestashop l'a voulu par sécurité !
